### ISSUE TITLE
[BUG] QNGOptimizer with molecular_hamiltonian

### ISSUE THREAD

**Issue by KetpuntoG** [PennyLaneAI/pennylane#2502]
### Expected behavior

The following code was to be executed:

```python

import pennylane as qml
from pennylane import qchem
from pennylane import numpy as np
import time

symbols = ["Li", "H"]
geometry = np.array([0.0, 0.0, 0.0, 0.0, 0.0, 2.969280527])

H, qubits = qchem.molecular_hamiltonian(
    symbols,
    geometry,
    active_electrons=2,
    active_orbitals=5
)

dev = qml.device("default.qubit", wires = 10)

@qml.qnode(dev)
def circuit(param):
    qml.DoubleExcitation(param[0], wires=[0,1,2,3])
    return qml.expval(H)
       
opt = qml.QNGOptimizer(stepsize=0.5)

param = np.array([1.], requires_grad = True)
param = opt.step(circuit, param)
print(param)

```

### Actual behavior

the code does not work because this error is displayed:

```
DiagGatesUndefinedError
```

### Additional information

The above error is due to the fact that for some reason the DoubleExcitation operator is not expanded in the process. Decomposing manually solves the problem.

```python
import pennylane as qml
from pennylane import qchem
from pennylane import numpy as np
import time

symbols = ["Li", "H"]
geometry = np.array([0.0, 0.0, 0.0, 0.0, 0.0, 2.969280527])

H, qubits = qchem.molecular_hamiltonian(
    symbols,
    geometry,
    active_electrons=2,
    active_orbitals=5
)

dev = qml.device("default.qubit", wires = 10)

@qml.qnode(dev)
def circuit(param):
    with qml.tape.stop_recording():
        ops = qml.DoubleExcitation(param[0], wires=[0,1,2,3]).decomposition()
    for op in ops:
        qml.apply(op)
    return qml.expval(H)
       
opt = qml.QNGOptimizer(stepsize=0.5)

param = np.array([1.], requires_grad = True)
param = opt.step(circuit, param)
print(param)
```

However, a new error appears:

```
Operation Hamiltonian is not written in terms of a single parameter
```

This error is due to casting when obtaining the qchem Hamiltonian. It can be solved manually by adding the following line after defining the Hamiltonian:

```python
H = qml.Hamiltonian([float(co) for co in H.coeffs], H.ops)
```

But this is something that should be fixed internally and not have to be done manually.

### System information

```shell
Name: PennyLane
Version: 0.24.0.dev0
Summary: PennyLane is a Python quantum machine learning library by Xanadu Inc.
Home-page: https://github.com/XanaduAI/pennylane
Author: None
Author-email: None
License: Apache License 2.0
Location: /Users/guille/Documents/GitHub/pennylane
Requires: numpy, scipy, networkx, retworkx, autograd, toml, appdirs, semantic-version, autoray, cachetools, pennylane-lightning
Required-by: PennyLane-Qchem, PennyLane-Lightning, PennyLane-SF, PennyLane-qiskit, my-research

Platform info:           macOS-10.16-x86_64-i386-64bit
Python version:          3.8.8
Numpy version:           1.20.2
Scipy version:           1.6.2
Installed devices:
- lightning.qubit (PennyLane-Lightning-0.22.1)
- strawberryfields.fock (PennyLane-SF-0.20.0.dev0)
- strawberryfields.gaussian (PennyLane-SF-0.20.0.dev0)
- strawberryfields.gbs (PennyLane-SF-0.20.0.dev0)
- strawberryfields.remote (PennyLane-SF-0.20.0.dev0)
- strawberryfields.tf (PennyLane-SF-0.20.0.dev0)
- qiskit.aer (PennyLane-qiskit-0.19.0.dev0)
- qiskit.basicaer (PennyLane-qiskit-0.19.0.dev0)
- qiskit.ibmq (PennyLane-qiskit-0.19.0.dev0)
- qiskit.ibmq.circuitrunner (PennyLane-qiskit-0.19.0.dev0)
- qiskit.ibmq.sampler (PennyLane-qiskit-0.19.0.dev0)
- default.gaussian (PennyLane-0.24.0.dev0)
- default.mixed (PennyLane-0.24.0.dev0)
- default.qubit (PennyLane-0.24.0.dev0)
- default.qubit.autograd (PennyLane-0.24.0.dev0)
- default.qubit.jax (PennyLane-0.24.0.dev0)
- default.qubit.tf (PennyLane-0.24.0.dev0)
- default.qubit.torch (PennyLane-0.24.0.dev0)
```

### Existing GitHub issues

- [X] I have searched existing GitHub issues to make sure the issue does not already exist.

**Comment by antalszava** [comment]
Thank you @KetpuntoG for the report! @soranjh any thoughts here?

> However, a new error appears:
Operation Hamiltonian is not written in terms of a single parameter
This error is due to casting when obtaining the qchem Hamiltonian. It can be solved manually by adding the following line after defining the Hamiltonian:
H = qml.Hamiltonian([float(co) for co in H.coeffs], H.ops)
But this is something that should be fixed internally and not have to be done manually.

Would this error be resolved even if `requires_grad=False` was set for the coefficients? We were discussing this with Soran and came to the conclusion, that `requires_grad=False` would have to be set by users explicitly because we would like them to have the mindset of using the differentiable HF solver.

**Comment by antalszava** [comment]
Having said that, some more pointers or a dedicated error message could be added to help users here.

**Comment by KetpuntoG** [comment]
> 

I tried that:
```python
geometry = np.array([0.0, 0.0, 0.0, 0.0, 0.0, 2.969280527], requires_grad = False)
```

and the error is the same

**Comment by soranjh** [comment]
The bug is not particular to Hamiltonians generated with the `molecular_hamiltonian` function. 

1. `QNGOptimizer` fails even if the Hamiltonian is `H = qml.Hamiltonian([1.0], [qml.Identity(wires=[0])])` when the circuit contains `qml.DoubleExcitation(param[0], wires=[0,1,2,3])`. 

2. `QNGOptimizer` fails with any Hamiltonian that has differentiable coefficients, i.e., where `H.coeffs` has `requires_grad=True`. 

In the [Accelerating VQEs with quantum natural gradient](https://pennylane.ai/qml/demos/tutorial_vqe_qng.html#hydrogen-vqe-example) demo a solution is provided for the second issue:

> We also need to make the Hamiltonian coefficients non-differentiable by setting requires_grad=False.
> `hamiltonian = qml.Hamiltonian(np.array(hamiltonian.coeffs, requires_grad=False), hamiltonian.ops)`

Another solution could be modifying the `QNGOptimizer` to handle differentiable Hamiltonians properly. As another solution, the Hamiltonian coefficients can be made non-differentiable by default, which is not the best option from the qchem perspective.

**Comment by KetpuntoG** [comment]
Okay, I'll try to make some time today to solve the problem. Thanks! 👌

**Comment by antalszava** [comment]
Hi @KetpuntoG just double-checking: did #2524 only partially solve this issue?

**Comment by KetpuntoG** [comment]
One issue is solved and the other is solved if you use:
```python
hamiltonian = qml.Hamiltonian(np.array(hamiltonian.coeffs, requires_grad=False), hamiltonian.ops)
```
So it is not an error as such. It's just to keep in mind that the Hamiltonian has trainable parameters by default and we should set `requires_grad = False`.

**Comment by antalszava** [comment]
Thanks, closing this.
